### ISSUE TITLE
macOS: Remote operations fail with permission denied 

### ISSUE THREAD

**Issue by kyberdin** [sublimehq/sublime_merge#903]
**Version info**

 - OS: macOS 10.15.6
 - Build: 2032

**Description**
This is related to #608 .

All remote operations fail to access the remote repository, even though SSH configurations look OK and the git command executed by Sublime Merge runs fine in a terminal. I have to use my terminal to do any remote interaction with my repositories; Sublime Merge cannot access any of them even with keys loaded into an agent and keychain.

Sublime Merge's console states that this is what gets executed:

```
git -c credential.helper=/Users/kyberdin/Applications/Sublime\ Merge.app/Contents/MacOS/git-credential-sublime fetch origin --progress
```

This fails through the UI with exit code `128`.

Running this same command in a terminal in the same repository succeeds without issue (with an agent and keychain containing the appropriate SSH key).

**Steps to reproduce**

1. Add SSH key to agent (and keychain on Mac with `-K`)
2. Check remote access on terminal via `git fetch`
3. Open repository in Sublime Merge
4. Check remote access by performing a fetch through the pull dropdown
<img width="496" alt="sublimemerge_fetch_failure" src="https://user-images.githubusercontent.com/19895583/92310176-086a2500-ef69-11ea-9f66-cb34e3fb1077.png">

**Expected behavior**

Fetch, push, and pull access the repository without issue.

**Notes**

Example of my `~/.ssh/config`:

```
Host github
    Hostname github.com
    User git
    AddKeysToAgent yes
    UseKeychain yes
    PreferredAuthentications publickey
    IdentityFile /Users/kyberdin/.ssh/<path to my key>
```

* I have tried installing Sublime Merge in `/Applications` and `~/Applications`, but the issue persists.
* I have verified that my SSH key is in Keychain
* Fails using system `git` and Homebrew installation of `git`

**Comment by dpjohnst** [comment]
Hi @kyberdin,

Thank you for reporting this issue, and sorry to hear you're having trouble using Sublime Merge.
To investigate further, could you run the following commands from the terminal (in the repository) and share the output:
1. `git credential-osxkeychain`
2. `git config credential.helper --show-scope`

Also, could you share your debug information from Sublime Merge?
1. Open Sublime Merge
2. Open the repository where the issue occurs
3. Navigate to and select `Help > Debug Information` in the menu bar
4. The debug information will now be copied to the clipboard.Take a moment to review the debug information, and remove any information you feel is sensitive (email address, file path, remote repository addresses etc).
5. Share this information here.

Thanks,
\- Dylan

**Comment by kyberdin** [comment]
Hi @dpjohnst, thanks for following up on this.

As requested:

```
$ git credential-osxkeychain
usage: git credential-osxkeychain <get|store|erase>
```

Tried with `get`, and my terminal seemed to hang until I `CTRL+C`'d it.

<br>

```
$ git config --show-scope credential.helper
system	osxkeychain
```

Debug information:
```
=== App Version Information ===
Build: 2032

=== Git Version Information ===
Using Git: /Users/kyberdin/homebrew/Cellar/git/2.28.0/bin/git (custom)
git version 2.28.0
PATH: /Users/kyberdin/bin:
/Users/kyberdin/Library/Python/3.8/bin:
/Users/kyberdin/.cargo/bin:
/Users/kyberdin/homebrew/bin:
/Users/kyberdin/homebrew/Cellar/openjdk/latest/bin:
/usr/local/bin:
/usr/bin:/bin:/usr/sbin:/sbin:
/opt/X11/bin (from shell)
environment variables loaded using: /bin/zsh -l

=== Browse Page Information ===
HEAD: 614df4e3243655acfcb07f8a6181ccd18d857ec1
Is in merge: 0
Is in cherry_pick: 0
Is in rebase: 0
Is in revert: 0

=== Git Status Information ===
?? sublime/syntax/

=== Our Status Information ===
?? sublime/syntax/Make_Output.sublime-syntax

=== Git Config Information ===
alias.a=add
alias.b=branch
alias.bl=blame
alias.ch=checkout
alias.cm=commit -m
alias.com=commit
alias.cp=cherry-pick
alias.drop=!sh -c "git stash drop stash@{$1};"
alias.pl=pull
alias.pop=!sh -c "git stash pop stash@{$1};"
alias.pu=push
alias.rb=rebase
alias.s=stash
alias.shows=!sh -c "git show -p stash@{$1};"
alias.st=status
branch.master.merge=refs/heads/master
branch.master.rebase=true
branch.master.remote=origin
core.bare=false
core.editor=vim
core.eol=lf
core.filemode=true
core.ignorecase=true
core.logallrefupdates=true
core.mergeoptions=--no-edit
core.precomposeunicode=true
core.repositoryformatversion=0
core.whitespace=cr-at-eol
credential.helper=osxkeychain
merge.tool=vim
mergetool.vim.path=/usr/bin/vim
push.default=simple
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
remote.origin.url=git@gitlab.com:kyberdin/configs.git
user.email=kyberdin.git@gmail.com
user.name=kyberdin

=== Our Config Information ===
Git Config Path Information
Using config path: /Users/kyberdin/homebrew/Cellar/git/2.28.0/etc/gitconfig
Using config path: /Users/kyberdin/.config/git/config
Using config path: /Users/kyberdin/.gitconfig
Using config path: /Users/kyberdin/configs/.git/config

Our config output is different to the Git config output
alias.a=add
alias.b=branch
alias.bl=blame
alias.ch=checkout
alias.cm=commit -m
alias.com=commit
alias.cp=cherry-pick
alias.drop=!sh -c "git stash drop stash@{$1};"
alias.pl=pull
alias.pop=!sh -c "git stash pop stash@{$1};"
alias.pu=push
alias.rb=rebase
alias.s=stash
alias.shows=!sh -c "git show -p stash@{$1};"
alias.st=status
branch.master.merge=refs/heads/master
branch.master.rebase=true
branch.master.remote=origin
core.bare=false
core.editor=vim
core.eol=lf
core.filemode=true
core.ignorecase=true
core.logallrefupdates=true
core.mergeoptions=--no-edit
core.precomposeunicode=true
core.repositoryformatversion=0
core.whitespace=cr-at-eol
merge.tool=vim
mergetool.vim.path=/usr/bin/vim
push.default=simple
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
remote.origin.url=git@gitlab.com:kyberdin/configs.git
user.email=kyberdin.git@gmail.com
user.name=kyberdin

=== Git Attributes Information ===
git check_attr --all output

=== Our Modified Files Newline Normalisation and EOL Information ===

=== Our Modified Files Flag Information ===
Ignoring symlinks: 0
```

**Comment by dpjohnst** [comment]
hi @kyberdin,

Thanks for sharing this - it looks as if Sublime Merge is not picking up the same config files as the version of Git used on the command line.

Could you try run the following command and confirm whether the issue persists:
`git config --global credential.helper osxkeychain`

Thanks,
\- Dylan

**Comment by kyberdin** [comment]
Hi @dpjohnst , very sorry for taking so long to get back to you about this.

Yes, it looks like the problem still persists. I did notice this change in the debug output, however:

```
Our config output is identical to the Git config output
```

**Comment by kyberdin** [comment]
Hi @dpjohnst , has there been any progress on this? I noticed it still persists in the most recent 2038 release.

**Comment by dpjohnst** [comment]
Hi @kyberdin,

Thank you for your patience while we investigate this.
Unfortunately we haven't been able to replicate the issue so far. 

That being said, these kinds of issues can sometimes be caused by restrictive permissions on the private key.
Could you confirm the permissions set on the private key file (run `ls -l /path/to/file` via the terminal)?

Kind regards,
\- Dylan

**Comment by dpjohnst** [comment]
Hi @kyberdin,

It looks like you may have two OpenSSH installations - the system version and the homebrew version.
You're likely seeing a conflict between the two, with Merge using the system version, and the terminal using the homebrew version.

Unfortunately I don't have access to a MacOS device currently, but I'd say your two options are:
* Uninstall the homebrew version
* Configure the homebrew version to fully integrate with OSX
This may help with integration: https://coderwall.com/p/qdwcpg/using-the-latest-ssh-from-homebrew-on-osx

Kind regards,
\- Dylan

**Comment by kyberdin** [comment]
Hi @dpjohnst ,

Your last comment got me thinking and, after some digging, I found the issue. There was another program running that interacts with ssh agents, and it was overwriting `SSH_AUTH_SOCK` in the terminal with a different sock path (as seen in my last post). So the keys I added were available in the terminal, but were not available for Merge.

I've used this program for years on Linux without any wonky SSH behavior, so it didn't register with me that it could be the culprit. The Linux version of Merge doesn't run into this problem with the program either, so there must be some differences for the program's MacOS version.

Either way, it's resolvable on my end and this issue can be closed as not a bug. Thank you very much for your time and assistance with this!
